protocol/communities/manager.go

```diff
@@ -5153,3 +5153,92 @@ func (c *Community) ToStatusLinkPreview() (*common.StatusCommunityLinkPreview, e
 
 	return communityLinkPreview, nil
 }
+
+func (m *Manager) determineChannelsForHRKeysRequest(c *Community, now int64) ([]string, error) {
+	result := []string{}
+
+	channelsWithMissingKeys := func() map[string]struct{} {
+		r := map[string]struct{}{}
+		for id := range c.Chats() {
+			if c.HasMissingEncryptionKey(id) {
+				r[id] = struct{}{}
+			}
+		}
+		return r
+	}()
+
+	if len(channelsWithMissingKeys) == 0 {
+		return result, nil
+	}
+
+	requests, err := m.persistence.GetEncryptionKeyRequests(c.ID(), channelsWithMissingKeys)
+	if err != nil {
+		return nil, err
+	}
+
+	for channelID := range channelsWithMissingKeys {
+		request, ok := requests[channelID]
+		if !ok {
+			// If there's no prior request, ask for encryption key now
+			result = append(result, channelID)
+			continue
+		}
+
+		// Exponential backoff formula: initial delay * 2^(requestCount - 1)
```

Comment:
Do you think it might be good to have that code out of the `determineChannelsForHRKeysRequest`, let's say in it's own package if we wanna be perfectionists 😄 

- For instance suppose we want to add a `jitter` to the exponential backoff, then this function will start to grow etc.
- Other parts of the code could benefit from this code as well
- Changing the retry strategy could also be easier if out of this `determineChannelsForHRKeysRequest` function

Reply:
Thanks, all the concerns are valid, and I could, in theory, extract the `NextBackOff` function into a separate package, but that feels like overkill to me. Additionally, I’m pretty sure it will only be used here, making it a classic case of YAGNI.

```
func NextBackOff(initialDelay, startTime int64, attemptCount int) int64 {
    backoffDuration := initialDelay * (1 << (attemptCount - 1))
    nextTime := startTime + backoffDuration
    return nextTime
}
```